[TX Builder] Added react-hook-form to handle forms

DaniSomoza opened the pull request:
## What it solves
Resolves #238

## How this PR fixes it

We added [`react-hook-form`](https://react-hook-form.com/) to handle form state and validations.

We removed de [`Builder.tsx`](https://github.com/gnosis/safe-react-apps/pull/310/files#diff-16e5843f84a6e1b9db9232230df37ba17898969b8f316153ba26767187885567) component that handled the form state and errors using [React State](https://reactjs.org/docs/hooks-state.html).

To reduce code complexity and a lot of side effects in the form state and errors we create a new set of reusable components to easily create new forms (see [the new TX-Builder designs](https://www.figma.com/proto/G7snLqBXWNPANcOj5MhiR7/Transaction-Builder?page-id=1%3A2&node-id=55%3A17805&viewport=241%2C48%2C0.18&scaling=min-zoom&starting-point-node-id=55%3A16961) like [the new Update Transaction](https://www.figma.com/proto/G7snLqBXWNPANcOj5MhiR7/Transaction-Builder?page-id=1%3A2&node-id=109%3A5609&viewport=241%2C48%2C0.18&scaling=min-zoom&starting-point-node-id=55%3A16961) modal)

New Components:

- [`AddNewTransactionForm`](https://github.com/gnosis/safe-react-apps/pull/310/files#diff-1d8b6836108a9e277a5f0b306e8e8a202d66c1410136c7955d3eeb325a71ab2b): component with the implementation of the Form that allows you to create a new Transaction.
- [`SolidityForm`](https://github.com/gnosis/safe-react-apps/pull/310/files#diff-599c064c512140d431459828e4bd98145c0e500af4ea66622204e3f2e325cbb1): Reusable component to easily create a form with all solidity fields
- [`Field.tsx`](https://github.com/gnosis/safe-react-apps/pull/310/files#diff-09565c6fb700bd6bd27224ce1638abc0777340fa3e1263b6e8514dc37a00d65c): As we are using Controlled components, we need to use [the Controller](https://react-hook-form.com/api/usecontroller/controller#main) component [see react-hook-form docs](https://react-hook-form.com/api/usecontroller/controller#main) for more details.
- Other form field Components: `AddressContractField.tsx`, `SelectContractField.tsx`, `TextContractField.tsx`, `TextareaContractField.tsx`...
- Validations: `basicSolidityValidation.ts`, `validateAddressField.ts`, `validateBooleanField.ts` ...

Yo can add custom validations to a field type in `validateField.ts` file:
```
const CUSTOM_VALIDATIONS: CustomValidationsType = {
  [ADDRESS_FIELD_TYPE]: [validateAddressField],
  [AMOUNT_FIELD_TYPE]: [validateAmountField, uintBasicValidation],
  ...
  [YOUR_FIELD_TYPE]: [isEven, lessThan30, moreThan10...],
};
```

## How to test it

### Transfer ether Tx

1. We Enter a EOA address in the first address input (like `rin:0x680cde08860141F9D223cE4E620B10Cd6741037E`)
2. To Address field should be auto-filled with the EOA used in the previous step.
3. ETH Value field is required (validation should be trigged if user clicks on "Add Transaction" Button).
4. If a valid EOA address and  a valid ETH amount is provided, the transaction should be added to the transaction list if user clicks on "Add Transaction" Button:
![Captura de pantalla 2022-02-03 a las 18 14 22](https://user-images.githubusercontent.com/26763673/152394102-68e39af3-0c2d-47d8-a36d-162a0351c71b.png)

### Data Hex Encoded Field

Use custom data (hex encoded)

1. User can send custom Hex data to a valid address using the Use custom data switch button
2. If a non hex value is provided an error should be present in the textarea field.
3.  If a valid hex value is provided the transaction should be added to the transaction list:
![Captura de pantalla 2022-02-03 a las 18 20 16](https://user-images.githubusercontent.com/26763673/152394780-494b898d-a100-4fd2-ab15-cea6277d3fce.png)

### Smart contract Interaction using an address:

1. We Enter a Contract address in the first address input: `0x49d4450977E2c95362C13D3a31a09311E0Ea26A6`
2. Select a Method in the selector (`testAddress` method)
3. fill `paramAddress` (address field) with a valid address `0x680cde08860141F9D223cE4E620B10Cd6741037E`
4. Click on "Add Transactions.
![Captura de pantalla 2022-02-04 a las 11 03 19](https://user-images.githubusercontent.com/26763673/152510152-8bbb0901-453f-4d07-b8cf-1471c7e83de6.png)

### Smart contract Interaction using ABI:
You can use this ABI:
```
[{"inputs":[{"internalType":"address","name":"paramAddress","type":"address"}],"name":"testAddress","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"address[]","name":"paramArrayAddress","type":"address[]"}],"name":"testArrayAddress","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bool[]","name":"paramArrayBool","type":"bool[]"}],"name":"testArrayBool","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bytes[]","name":"paramArrayBytes","type":"bytes[]"}],"name":"testArrayBytes","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bytes[]","name":"paramArrayBytes32","type":"bytes[]"}],"name":"testArrayBytes32","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256[]","name":"paramArrayInteger","type":"uint256[]"}],"name":"testArrayInteger","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256[]","name":"","type":"uint256[]"}],"name":"testArrayNoName","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bool","name":"paramBool","type":"bool"}],"name":"testBool","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bool","name":"","type":"bool"}],"name":"testBoolNoName","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bytes","name":"paramBytes","type":"bytes"}],"name":"testBytes","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bytes32","name":"paramBytes16","type":"bytes32"}],"name":"testBytes16","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bytes32","name":"paramBytes32","type":"bytes32"}],"name":"testBytes32","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"paramInteger","type":"uint256"}],"name":"testInteger","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint8","name":"paramInteger16","type":"uint8"}],"name":"testInteger16","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint8","name":"paramInteger16","type":"uint8"}],"name":"testInteger8","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"paramLongName_abcdefghijklmnopqrstuvwxyz0123456789abcdefghijklmnopqrstuvwxyz0123456789","type":"uint256"}],"name":"testIntegerLongParameterName","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"","type":"uint256"}],"name":"testIntegerNoName","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"p1","type":"uint256"},{"internalType":"bool","name":"p2","type":"bool"},{"internalType":"bytes32","name":"p3","type":"bytes32"},{"internalType":"address","name":"p4","type":"address"},{"internalType":"bool[][]","name":"p5","type":"bool[][]"},{"internalType":"bytes32[]","name":"p6","type":"bytes32[]"},{"internalType":"address[]","name":"p7","type":"address[]"}],"name":"testManyParams","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"p1","type":"uint256"},{"internalType":"bool","name":"","type":"bool"},{"internalType":"bytes32","name":"","type":"bytes32"},{"internalType":"address","name":"p4","type":"address"},{"internalType":"uint256[]","name":"","type":"uint256[]"},{"internalType":"bool[][]","name":"","type":"bool[][]"},{"internalType":"bytes32[]","name":"p7","type":"bytes32[]"},{"internalType":"address[]","name":"p8","type":"address[]"}],"name":"testManyParamsMix","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256","name":"","type":"uint256"},{"internalType":"bool","name":"","type":"bool"},{"internalType":"bytes32","name":"","type":"bytes32"},{"internalType":"address","name":"","type":"address"},{"internalType":"uint256[]","name":"","type":"uint256[]"},{"internalType":"bool[][]","name":"","type":"bool[][]"},{"internalType":"bytes32[]","name":"","type":"bytes32[]"},{"internalType":"address[]","name":"","type":"address[]"}],"name":"testManyParamsNoName","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bool[][][][]","name":"paramNestedArrayBool","type":"bool[][][][]"}],"name":"testNestedArrayBool4Levels","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"uint256[][]","name":"paramNestedArrayInteger","type":"uint256[][]"}],"name":"testNestedArrayInteger","outputs":[],"stateMutability":"nonpayable","type":"function"},{"inputs":[{"internalType":"bool[][][][][][][][][][]","name":"paramNestedArrayBool","type":"bool[][][][][][][][][][]"}],"name":"testNestedArrayInteger10Levels","outputs":[],"stateMutability":"nonpayable","type":"function"}]
```
1. We Enter the ABI in the first input:
2. Fill a ToAddress input: `0x680cde08860141F9D223cE4E620B10Cd6741037E`
3. Select a Method in the selector (`testAddress` method)
4. fill `paramAddress` (address field) with a valid address `0x680cde08860141F9D223cE4E620B10Cd6741037E`
5. Click on "Add Transactions.
![Captura de pantalla 2022-02-04 a las 11 18 19](https://user-images.githubusercontent.com/26763673/152512149-ae139f8a-6487-41cf-b817-0f60419325fe.png)


## Screenshots

Errors in the form:

![Captura de pantalla 2022-02-03 a las 18 21 10](https://user-images.githubusercontent.com/26763673/152395168-804289c2-b029-4c91-b155-c22ef80a6536.png)
![Captura de pantalla 2022-02-03 a las 18 21 42](https://user-images.githubusercontent.com/26763673/152395176-1287f99e-45c8-4b4e-88d3-d520b14a7dea.png)
![Captura de pantalla 2022-02-03 a las 18 22 07](https://user-images.githubusercontent.com/26763673/152395180-be1bbdae-be06-4613-b4d4-4f4dfc840f84.png)
![Captura de pantalla 2022-02-04 a las 11 06 17](https://user-images.githubusercontent.com/26763673/152510580-a97689a3-a5cf-4c80-bf86-5782350456e3.png)

iamacook commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,202 @@
+  return (
+    <form id={id} onSubmit={handleSubmit(submitAndResetForm)} noValidate>
+      {/* To Address field */}
+      <Field
```
A lot of these props are used across multiple fields, e.g. `fullWidth`. Maybe you could export them into a reuseable component?

iamacook commented on apps/tx-builder/src/components/forms/fields/AddressContractField.tsx:
```diff
@@ -0,0 +1,50 @@
+import { AddressInput } from '@gnosis.pm/safe-react-components';
+
+function AddressContractField({
+  id,
```
You could spread a lot of these directly onto the component and not destructure them. It's not necessary but saves you a lot of hassle.

iamacook commented on apps/tx-builder/src/components/forms/fields/Field.tsx:
```diff
@@ -0,0 +1,103 @@
+          }}
+          render={({ field, fieldState }) => (
+            <Component
+              name={field.name}
```
Here's another example where you could just spread `{...field}`.

iamacook commented on apps/tx-builder/src/components/forms/fields/Field.tsx:
```diff
@@ -0,0 +1,103 @@
+          shouldUnregister={shouldUnregister}
+          rules={{
+            required: {
+              value: true,
```
Maybe it's best to take this from a prop so that it isn't always required? You can default it to true when you destructure it from the props though.

DaniSomoza replied:
Good Catch! All fields are required now, but as you mention it's best to take this from a prop, because in the future we could need it:
```
          rules={{
            required: {
              value: required,
              message: 'Required',
            },
            valida
```

iamacook commented on apps/tx-builder/src/utils.ts:
```diff
@@ -102,3 +90,47 @@ export const isValidAddress = (address: string | null) => {
+        return parseInputValue(contractField, cleanValue);
+      });
+      const abi = abiCoder as unknown; // a bug in the web3-eth-abi types
+      const hexEncondedData = (abi as AbiCoder).encodeFunctionCall(contractMethod as AbiItem, parsedValues);
```
Is it safe to cast these?

DaniSomoza replied:
yep it seems that its a bug related with the `web3-eth-abi` lib.

iamacook commented on apps/tx-builder/package.json:
```diff
@@ -8,6 +8,7 @@
     "@material-ui/core": "^4.12.3",
     "axios": "^0.24.0",
     "lodash": "^4.17.21",
+    "react-hook-form": "^7.25.3",
```
There are [developer tools](https://react-hook-form.com/dev-tools) that you could install as a developer dependancy (rendering them only when you're in the developer env).

DaniSomoza replied:
done! :D

iamacook commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,202 @@
+    mode: 'onTouched', // This option allows you to configure the validation strategy before the user submits the form
+  });
+
+  const contractMethodIndex = watch(CONTRACT_METHOD_INDEX_FIELD_NAME);
```
If you use [`useWatch`](https://react-hook-form.com/api/usewatch), it will isolate re-renders.

DaniSomoza replied:
mmm, I think that in this case we should use `watch()` right?

I think that [useWatch](https://react-hook-form.com/api/usewatch) its meant to be used in a nested component. useWatch will isolate re-render only that component and potentially result in better performance.

Also that component needs to be wrapped with a [FormProvider](https://react-hook-form.com/advanced-usage#FormProviderPerformance), it uses a [React's Context](https://reactjs.org/docs/context.html) API. 
It think that it only makes sense if the data is passed through a very complex components tree. In that case it saves you to having to pass props down manually at every level.

What is your opinion about that?

iamacook replied:
I suppose as you're using it within this component then it may not be necessary.

It requires `control`, which is likely the explanation for the context.

iamacook reviewed (commented):
Looks way cleaner! 🚀

DaniSomoza commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,202 @@
+
+  const { handleSubmit, control, setValue, watch, getValues, reset, clearErrors } = useForm<SolodityFormValues>({
+    defaultValues: initialValues,
+    mode: 'onTouched', // This option allows you to configure the validation strategy before the user submits the form
```
Good catch!
From [the docs](https://react-hook-form.com/api/useform#main):

- mode: `all`:  Validation will trigger on the blur and change events.
- mode: `onTouched`:  Validation will trigger on the first blur event. After that, it will trigger on every change event.

Basically, I used `onTouched` mode because when you start typing, the error doesn't show until you blur the field for first time (after this, validation is triggered on every change field event)

But I will update this to mode: `all` because its the current behaviour in the tx-builder.

iamacook replied:
It's up to you. The comment seemed as though it was copied from an example.

yagopv replied:
For me personally onTouched leads to a better UX @DaniSomoza

DaniSomoza replied:
restored mode to `onTouched`

DaniSomoza commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,202 @@
+        getAddressFromDomain={getAddressFromDomain}
+        networkPrefix={networkPrefix}
+        fieldType={ADDRESS_FIELD_TYPE}
+        control={control}
```
As I mention in [this other comment](https://github.com/gnosis/safe-react-apps/pull/310/files#r798587579). I think that [FormProvider](https://react-hook-form.com/advanced-usage#FormProviderPerformance) only makes sense if the data is passed through a very complex components tree. In that case it saves you to having to pass props down manually at every level. 

But I think that it is not our case right?

yagopv replied:
If you add a Context API provider you lose the nice of the uncontrolled components don't you?

iamacook replied:
RHF provides hooks to interact with the `FormProvider` context and reduce associated re-renders.

iamacook replied:
If we're not nesting fields then we need not use it though.

DaniSomoza commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,202 @@
+        fieldType={AMOUNT_FIELD_TYPE}
+        fullWidth
+        required
+        showField={isValueInputVisible}
```
Yes, I am using 2 props for this (I need to finish my PR description to specify this sorry :< )

- prop `showField`: This allows you to visually show/hide the component field
- prop `shouldUnregister`: This allows you to keep/remove the field value in the form state when you visually show/hide the component field. (we need this when we switch between encoding/decoding contract fields values to hex data)

iamacook replied:
I think `showField` shoul be outside of the component, no? `flag && <Field {...props />`

You can enable `shouldUnregister` by default in `userForm` if it helps. Just a suggestion ;)

DaniSomoza replied:
I added a `showField` prop to perform the conditional rendering inside of the `Field` Component:
 
And regarding `shouldUnregister`, based [on the v7 docs](https://react-hook-form.com/api/useform): 
By default, an input value will be retained in the form state when input is removed. You can set `shouldUnregister` to `true` to unregister input during unmount in the `useForm`, but the problem is:
```
This is a global config that overwrites child-level config, if you want to have individual behavior, then you should set the config at the component or hook level, not at useForm.
```
In some cases we want to keep some values in the form state. Based on the docs the only way to do this properly is using directly the child-level config (using the `shouldUnregister` prop)

DaniSomoza replied:
No reason behind that, I removed the `showField`  prop from `Field` component.

now, if you want to show/hide a form field you shoould use conditional rendering from outside:

```
        {showContractFields && (
          <Field
            id="contract-method-selector"
            [...]
```

iamacook commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,202 @@
+        label={'Contract Method Selector'}
+        fieldType={CONTRACT_METHOD_FIELD_TYPE}
+        showField={showContractFields}
+        shouldUnregister={false}
```
Seems like the `useForm` argument overwrite child-level setting, so I'll resolve this.

DaniSomoza commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,202 @@
+      <Field
+        id="contract-method-selector"
+        name={CONTRACT_METHOD_INDEX_FIELD_NAME}
+        label={'Contract Method Selector'}
```
done!

DaniSomoza commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,202 @@
+            fieldType={contractField.type}
+            fullWidth
+            required
+            shouldUnregister={false} // required to keep contract field values in the form state when the user switches between encoding and decoding data
```
see [this comment](https://github.com/gnosis/safe-react-apps/pull/310#discussion_r800792700)

DaniSomoza commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,208 @@
+      </form>
+
+      {/* set up the dev tool */}
+      <DevTool control={control} />
```
done!

iamacook reviewed (approved):
Great work dude! 💪

dasanra commented on apps/tx-builder/src/components/forms/validations/validateAmountField.ts:
```diff
@@ -0,0 +1,11 @@
+import { ValidateResult } from 'react-hook-form';
+
+import { isInputValueValid } from '../../../utils';
```
Not to be done in this PR, but is better if we use the same approach of absolute routes if possible, we should check it

JagoFigueroa commented:
For now only issue I've found is that when working with assets, no decimals are allowed. @DaniSomoza is already doing his magic to fix it

<img width="553" alt="Screenshot 2022-02-14 at 16 26 23" src="https://user-images.githubusercontent.com/92332710/153894711-e5baf26f-60bb-43b5-bddd-497171d46b03.png">

Cheers!

DaniSomoza commented:
@JagoFigueroa  Good Catch! 

Fixed! :D

I added this validation to the `validateAmountField.ts` fn:

```
  // should be a valid amount in wei
  try {
    toWei(value);
  } catch (error) {
    return INVALID_AMOUNT_ERROR;
  }
```

I added a unit test to make sure that we show errors for invalid decimal amounts:

```
      it('validates invalid decimal amounts', () => {
        const validationResult = validateAmountField('0.000000000000000000000000000000001');

        expect(validationResult).toBe('Invalid amount value');
      });
```

mikheevm commented on apps/tx-builder/src/components/forms/AddNewTransactionForm.tsx:
```diff
@@ -0,0 +1,137 @@
+    [CONTRACT_METHOD_INDEX_FIELD_NAME]: '0',
+  };
+
+  function onSubmit(values: SolidityFormValuesTypes) {
```
this function is defined with `function` keyword but the component is defined with an arrow function, let's stay consistent

DaniSomoza replied:
done!

mikheevm commented on apps/tx-builder/src/components/forms/SolidityForm.tsx:
```diff
@@ -0,0 +1,214 @@
+
+  const isValueInputVisible = showHexEncodedData || !contract || isPayableMethod;
+
+  function onClickShowHexEncodedData(checked: boolean) {
```
same here with the definitions

DaniSomoza replied:
done!

mikheevm commented on apps/tx-builder/src/components/forms/fields/TextareaContractField.tsx:
```diff
@@ -0,0 +1,10 @@
+const DEFAULT_ROWS = 4;
+
+const TextareaContractField = (props: TextFieldInputProps) => {
+  return <TextContractField {...props} multiline rows={DEFAULT_ROWS} />;
```
should rows be editable?

DaniSomoza replied:
Good point! 

Now `rows` are editable:

```
<TextContractField {...props} multiline rows={props.rows || DEFAULT_ROWS} />;
```

JagoFigueroa commented:
All bueno! Gracias @DaniSomoza :pepe-king:
